File: android/Girlsfleet/app/src/main/java/com/circus/girlsfleet/SurfaceActivity.java
package com.circus.girlsfleet;

import androidx.appcompat.app.AppCompatActivity;

import android.content.res.AssetManager;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.os.Bundle;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import com.circus.girlsfleet.fleet.MapState;

import java.util.Random;
import java.util.Stack;

public class SurfaceActivity extends AppCompatActivity implements IGameSystem,IRenderer,Runnable,SurfaceHolder.Callback {

    private volatile boolean m_isRunning = true;
    private Thread m_thread = null;
    private SurfaceHolder m_surfaceHolder = null;
    private Random m_rad = new Random();
    private ScreenMode m_screenMode;
    private Stack<IGameState> m_stateStack = new Stack<>();

    private Canvas m_canvas;
    private Paint m_paint;
    private int m_screenWidth,m_screenHeight;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_surface);

        SurfaceView sw = findViewById(R.id.ScreenSurface);
        m_surfaceHolder = sw.getHolder();
        m_surfaceHolder.addCallback(this);
    }

    public ScreenMode getScreenMode() {
        return m_screenMode;
    }

    public AssetManager getAssetManager(){
        return getAssets();
    }
    public Random getRandom(){
        return m_rad;
    }
    public Stack<IGameState> getStateStack(){
        return m_stateStack;
    }
    @Override
    public void onPause() {
        super.onPause();
        m_isRunning = false;
        if (null != m_thread) {
            try {
                m_thread.join();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public void surfaceCreated(SurfaceHolder holder) {

    }

    public void surfaceChanged(SurfaceHolder holder, int format, int w, int h) {
        m_screenMode = ScreenMode.HD_PLUS_VERTICAL;
        m_screenWidth=w;
        m_screenHeight=h;
        adjustSurfacePos(w,h);

        stopGameThread();
        m_isRunning = true;
       // m_stateStack.push(new AzurelaneMapState(this, AzurelaneMapLoader.load("1_1.json",getAssets())));
         m_stateStack.push(new MapState(this, MapLoader.load("1_1.json",getAssets())));
        startGameThread();
    }

    private void adjustSurfacePos(int w,int h) {
        int lOffset=(w-m_screenMode.getRect().width())/2;
        int tOffset=(h-m_screenMode.getRect().height())/2;
        m_screenMode.getRect().offsetTo(lOffset,tOffset);

    }

    private void stopGameThread() {
        if (m_isRunning && null!=m_thread) {
            try {
                m_isRunning = false;
                m_thread.join();
            } catch (Exception e) {
                e.printStackTrace();
            }

        }

    }

    private void startGameThread() {
        m_isRunning = true;
        m_thread = new Thread(this);
        m_thread.start();

    }

    public void surfaceDestroyed(SurfaceHolder holder) {

    }

    public void run() {
        long prevFrameTime = 0L;
        long prevUpdateTime = 0L;
        m_paint = new Paint();
        while (m_isRunning) {
            //Log.i(this.getClass().getName(), String.format("%d states in stack", m_stateStack.size()));
            if (0 < m_stateStack.size()) {
                //Log.i(this.getClass().getName(), String.format("Top stack is %s", m_stateStack.peek().toString()));
                m_stateStack.peek().onUpdate(this);
            }

            if (0 == m_stateStack.size()) {
                Log.i(this.getClass().getName(), "Ending the main loop");
                m_isRunning = false;
                break;
            }
            long curTime = System.currentTimeMillis();
            if (FPSHelper.waitFrame(prevFrameTime, prevUpdateTime, curTime)) {
                m_canvas = m_surfaceHolder.lockCanvas();
                setColor(Color.BLACK);
                fillRect(0, 0, m_screenWidth,m_screenHeight);
                if (0 < m_stateStack.size()) {
                    m_stateStack.peek().onDraw(this);
                }

                m_surfaceHolder.unlockCanvasAndPost(m_canvas);
                prevUpdateTime = curTime;

            }
            prevFrameTime = curTime;
        }

        m_canvas = null;
        m_paint = null;
        if (!m_isRunning && m_stateStack.empty()) {
            Log.i(this.getClass().getName(), "Terminate the activity");
            this.finish();
        }

    }


    public void drawRect(Rect r) {
        m_paint.setStyle(Paint.Style.STROKE);
        m_canvas.drawRect(r, m_paint);
    }

    public void drawRect(int l, int t, int w, int h) {
        m_paint.setStyle(Paint.Style.STROKE);
        m_canvas.drawRect(l, t, l + w, t + h, m_paint);
    }

    public void fillRect(Rect r) {
        m_paint.setStyle(Paint.Style.FILL);
        m_canvas.drawRect(r, m_paint);
    }

    public void fillRect(int l, int t, int w, int h) {

        m_paint.setStyle(Paint.Style.FILL);
        m_canvas.drawRect(l,t,l+w,t+h,m_paint);
    }
    public void fillRect(Vector lt,int radius){
        m_paint.setStyle(Paint.Style.FILL);
        m_canvas.drawRect(lt.getX()-radius,lt.getY()-radius,lt.getX()+radius,lt.getY()+radius,m_paint);

    }
    public void drawLine(int x1,int y1,int x2,int y2){
        m_canvas.drawLine(x1,y1,x2,y2,m_paint);
    }
    public void drawText(int l,int t,String txt) {
        m_canvas.drawText(txt,l,t,m_paint);
    }
    public void setColor(int c){
        m_paint.setColor(c);
    }
}

File: android/Circus/app/src/main/java/com/personal/circus/SurfaceActivity.kt
package com.personal.circus

import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.*
import java.util.*
import kotlin.collections.HashMap
import android.R
import android.graphics.Color
import android.util.Log
import com.personal.circus.Kancole.DungeonState
import com.personal.circus.Kancole.RootState

class SurfaceActivity : AppCompatActivity() ,IGameSystem,Runnable,SurfaceHolder.Callback {
    @Volatile
    private var m_isRunning = false
    private var m_thread: Thread? = null
    private var m_surfaceHolder: SurfaceHolder? = null
    private val m_stateStack = Stack<IGameState>()
    private val m_renderer=(CanvasRenderer() as IRenderer)
    private var m_rand: Random = Random();
    private var m_screenMode: ScreenMode? = null

    companion object{
        @JvmStatic
        val GAME_TYPE="GameType"
        val GAME_KANCORE="Kancore"
        val GAME_KANCORE_ROOT_BOX="KancoreRootBox"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (Build.VERSION.SDK_INT >= 19) {
            val decor = this.window.decorView
            decor.systemUiVisibility =
                View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_FULLSCREEN or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        } else {
            window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        }
        setContentView(com.personal.circus.R.layout.activity_surface)

        //アクションバーを非表示
        val ab = supportActionBar
        ab!!.hide()


        //乱数器を生成
        m_rand.setSeed(System.currentTimeMillis())

        //SurfaceViewの初期化
        val sv = findViewById<SurfaceView>(com.personal.circus.R.id.GameSurface)

        m_surfaceHolder = sv.holder
        m_surfaceHolder?.addCallback(this)



    }



    /**
     *
     * @return javaデフォルトのRandオブジェクトを返す
     */
    override fun getRand(): Random {
        return m_rand
    }

    override fun getScreenMode(): ScreenMode {
        return m_screenMode!!
    }

    override fun getStateStack(): Stack<IGameState> {
        return m_stateStack
    }
    override fun onPause() {
        super.onPause()
        m_isRunning = false
        if(null!=m_thread) {
            try {
                m_thread!!.join()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

    }

    override fun onResume() {
        super.onResume()
//        m_isRunning = true
//                m_thread!!.start();
        if (Build.VERSION.SDK_INT >= 19) {
            val decor = this.window.decorView
            decor.systemUiVisibility =
                View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_FULLSCREEN or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        }
    }

    override fun surfaceCreated(holder: SurfaceHolder) {}
    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {

        m_screenMode = ScreenMode.HD_PLUS_VERTICAL
        m_screenMode!!.setScreenSize(width,height)
        Log.i(this.javaClass.name, String.format("The window has resized.width=%d,height=%d", width, height))

        if(m_isRunning){
            try{
                m_isRunning=false
                m_thread!!.join()
            }catch (e:java.lang.Exception){
                e.printStackTrace()
            }

        }

        m_isRunning=true
        if(intent.hasExtra(GAME_TYPE)){
            m_stateStack.clear()
            val type=intent.getStringExtra(GAME_TYPE)
            if(GAME_KANCORE.equals(type) ){
                m_stateStack.add(DungeonState(this))
            }else if(GAME_KANCORE_ROOT_BOX.equals(type)){
                m_stateStack.add(RootState(this))
            }
        }
        //ゲームスレッドを生成
        //アクティビティの構築が終わったら開始する
        m_thread = Thread(this)
        m_thread!!.start()

    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {}

    override fun run() {

        var prevFrameTime: Long = 0
        var prevUpdateTime: Long = 0
        while (m_isRunning) {
            Log.i(this.javaClass.name,String.format("%d states in stack",m_stateStack.size))
            if (0 < m_stateStack.size) {
                Log.i(this.javaClass.name,String.format("Top stack is %s",m_stateStack.peek().toString()))
                m_stateStack.peek().onUpdate(this)
            }

            if(0==m_stateStack.size){
                Log.i(this.javaClass.name,"Ending the main loop")
                m_isRunning=false
                break
            }
            val curTime = System.currentTimeMillis()
            if (FPSHelper.waitFrame(prevFrameTime, prevUpdateTime, curTime)) {
                val c = m_surfaceHolder!!.lockCanvas() ?: continue
                (m_renderer as CanvasRenderer).bindCanvas(c)
                m_renderer!!.setColor(Color.BLACK)
                m_renderer!!.fillRect(0, 0, m_screenMode!!.getScreenWidth(), m_screenMode!!.getScreenHeight())
                if (0 < m_stateStack.size) {
                    m_stateStack.peek().onDraw(m_renderer!!)
                }

                m_surfaceHolder!!.unlockCanvasAndPost(c)
                prevUpdateTime = curTime

            }
            prevFrameTime = curTime
        }

        if(!m_isRunning && m_stateStack.empty()) {
            Log.i(this.javaClass.name,"Terminate the activity")
            this.finish()
        }

    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (0 < m_stateStack.size && MotionEvent.ACTION_UP == event.action) {
            m_stateStack.peek().onTouch(this, event.x.toInt(), event.y.toInt())
        }
        return true
    }




}
